### PR TITLE
feat: zilliqa consensus data related to block 

### LINKED ISSUE #10701: Zilliqa: Consensus data for Block

### PR DESCRIPTION
**Pull request by fedor-ivn**
Closes #10701.

### Enhancements

- **New modules for consensus data processing**: 
  - Introduced dedicated modules within `EthereumJSONRPC` for parsing Zilliqa-specific consensus data retrieved from `eth_getBlockByNumber`/`eth_getBlockByHash` RPC calls:
    - `EthereumJSONRPC.Zilliqa.QuorumCertificate`
    - `EthereumJSONRPC.Zilliqa.AggregateQuorumCertificate`
    - `EthereumJSONRPC.Zilliqa.NestedQuorumCertificate`

- **Extend `EthereumJSONRPC.Block` for consensus data**:
  - Modified `EthereumJSONRPC.Block` to extract `quorumCertificate` and `aggregateQuorumCertificate` directly from RPC call responses.
  - ⚠️ **Note**: The JSON parsing logic deviates slightly from other methods in `EthereumJSONRPC.Block` due to structural differences. See this [comment](https://github.com/blockscout/blockscout/pull/10699/files#r1800752796) for an in-depth explanation.

- **Extend `EthereumJSONRPC.Blocks` struct**:
  - Added chain-type dependent fields to `EthereumJSONRPC.Blocks` to store Zilliqa consensus data fields.

- **Database schema update**:
  - Added new tables to store Zilliqa-related consensus data (quorum certificate and aggregate quorum certificate).
  - For a schema overview, please refer to this [`dbdiagram.io`](https://dbdiagram.io/d/66db1abaeef7e08f0eedf9f3) visualization.

    Zilliqa Database schema

- **Block fetcher**:
	- Added new import options in Block fetcher.

- **API enhancements**:
	- Enhanced the `/api/v2/blocks/{block_hash_or_number}` endpoint to include Zilliqa consensus data.

### Refactoring

- **Refactor `Explorer.Repo`**:
  - DRY configuration of chain type and feature-dependent repositories within `apps/explorer/lib/explorer/repo.ex`.

- **Codebase consistency updates**:
  - Minor formatting adjustments in `.github/workflows/config.yml`.
  - Renamed a module in `apps/ethereum_jsonrpc/mix.exs` to align with common naming scheme.
  - Sorted chain types alphabetically in `config/config_helper.exs`.

### API Changes

The `/api/v2/blocks/{block_hash_or_number}` endpoint now includes a new `zilliqa` field in the response. This field provides PBFT consensus details, such as:

- **`view`**: The iteration of the consensus round in which the block was proposed.
- **`quorum_certificate`**: Always present in the response, indicating standard consensus agreement.
- **`aggregate_quorum_certificate`**: Typically `null`. This field appears if validators timed out (when no block was proposed in the prior view). In these cases, the next leader includes an `aggregate_quorum_certificate` along with the highest QC to the new block.

#### For Block `1135006` (Standard Quorum Certificate Only):

```jsonc
{
    "zilliqa": {
        "view": 1137867,
        "quorum_certificate": {
            "view": 1137867,
            "signature": "0xa06f119e4d377a1664c5b5134cbfa180333d22b7b7a4e2e5a06175e9c904008ae187ae2470d4f71a6fc0ed9c1d70052f120efa56b866136755bbf70c89857f7e94e4b5624dd8b60ceb37bd8e1510f5694e2c95db06d44dc1c2c500e4fca4289c",
            "signers": [1, 2, 3, 8]
        },
        "aggregate_quorum_certificate": null
    },
    // other block fields
}
```

#### For Block `1135005` (With Aggregate Quorum Certificate):

```jsonc
{
    "zilliqa": {
        "view": 1137866,
        "quorum_certificate": {
            "view": 1137866,
            "signature": "0xa78c7f3e07e1df963ddeda17a1e5afd97c7c8a6fc8e0616249c22a2a1cc91f8eef6073cab8ba22b50cc7b38090f1ad9109473d30f24d57858d1f28c6679b3c4deeb800e5572b5e15604596594d506d3103a44d8b707da581f1a4b82310aeecb6",
            "signers": [0, 1, 3, 8]
        },
        "aggregate_quorum_certificate": {
            "view": 1137866,
            "signature": "0x820f591cd78b29a69ba25bc85c4327fa3b0adb61a73a4f0bd943b4ab0b97e061eae9ac032d19fbfab7efb89fac2454ab0b89fea83185c0dac749ff55b0e2c21535a2b712872491577728db868d11939461a6bfde0d94d238f46b643bbe19767e",
            "signers": [1, 2, 3, 8],
            "nested_quorum_certificates": [
                {
                    "view": 1137863,
                    "signature": "0xa78c7f3e07e1df963ddeda17a1e5afd97c7c8a6fc8e0616249c22a2a1cc91f8eef6073cab8ba22b50cc7b38090f1ad9109473d30f24d57858d1f28c6679b3c4deeb800e5572b5e15604596594d506d3103a44d8b707da581f1a4b82310aeecb6",
                    "signers": [0, 1, 3, 8],
                    "proposed_by_validator_index": 1
                },
                // other nested quorum certificates
            ]
        }
    },
    // other block fields
}
```

## Checklist for your Pull Request (PR)

- [x] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [x] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs). https://github.com/blockscout/docs/pull/329
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [x] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/block.ex [hunk 1]
```diff
@@ -872,6 +949,26 @@ defmodule EthereumJSONRPC.Block do
...
+
+      defp entry_to_elixir({"aggregateQuorumCertificate" = key, entry}, %{"hash" => block_hash}) do
+        # credo:disable-for-next-line Credo.Check.Design.AliasUsage
+        {key, EthereumJSONRPC.Zilliqa.AggregateQuorumCertificate.new(entry, block_hash)}
+      end
+
+    _ ->
+      :ok
+  end
+
```
Comment (fedor-ivn) [inline apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/block.ex:971]: As you may notice, I utilized slightly different approach to parsing consensus data. Here the raw entries are not parsed to just maps, but rather to module structs, which are then converted to params (`to_params/1` function) for insertion to DB.

File: apps/explorer/lib/explorer/repo.ex [hunk 2]
```diff
@@ -127,163 +127,35 @@ defmodule Explorer.Repo do
...
+    defmodule repo do
+      use Ecto.Repo,
+        otp_app: :explorer,
+        adapter: Ecto.Adapters.Postgres
+
+      def init(_, opts) do
+        ConfigHelper.init_repo_module(__MODULE__, opts)
+      end
     end
   end
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/repo.ex:161]: We can include here `Blacfort` as well, at the moment of development there was no such repo I suppose

File: config/runtime/dev.exs [hunk 3]
```diff
@@ -117,6 +117,15 @@ config :explorer, Explorer.Repo.PolygonZkevm,
   # separating repos for different CHAIN_TYPE is implemented only for the sake of keeping DB schema update relevant to the current chain type
   pool_size: 1
 
+# Configure Zilliqa database
+config :explorer, Explorer.Repo.Zilliqa,
```
Comment (sl1depengwyn) [inline config/runtime/dev.exs:121]: Have you tried declaring config in `for` simmilar to repo definitions?

**Review by sl1depengwyn — COMMENTED** [review]
We don't need fetcher for historic zilliqa consensus data?

**Comment by fedor-ivn** [comment]
> We don't need fetcher for historic zilliqa consensus data?

@sl1depengwyn As I got it, Zilliqa 2.0 mainnet will be completely new L1 network

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/zilliqa_view.ex [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+          nil
+
+        qc ->
+          zilliqa_json
+          |> Map.put(:quorum_certificate, %{
+            view: qc.view,
+            signature: qc.signature,
+            signers: qc.signers
+          })
+      end
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/block_scout_web/lib/block_scout_web/views/api/v2/zilliqa_view.ex:52]: Good catch! Thank you

File: config/runtime/dev.exs (hunk 3 above)
Comment (fedor-ivn) [reply config/runtime/dev.exs:128]: Yeah, I'll refactor other modules with `for` approach as well

File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/block.ex [hunk 5]
```diff
@@ -492,6 +533,7 @@ defmodule EthereumJSONRPC.Block do
 
   case Application.compile_env(:explorer, :chain_type) do
     :rsk ->
+      @spec chain_type_fields(params, elixir) :: params
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/block.ex]: For some reason, dialyzer began reporting unusual errors after I added this `@spec`. I spent about 1.5 hour investigating but couldn’t find the cause, so I just removed it.

[the remark this answers is not in the document]
Comment (sl1depengwyn) [reply apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/block.ex]: The reason behind the error is that `chain_type_fields` expects params with chain type fields, but it actually receives base params.

Possible solutions are:
- change `@spec chain_type_fields(params, elixir) :: params` to `@spec chain_type_fields(map(), elixir) :: params`
- change `@chain_type_fields` module attribute from 
```elixir
quote(
                           do: [
                             zilliqa_view: non_neg_integer()
                           ]
                         )
```
to
```elixir
quote(
                           do: [
                             {optional(:zilliqa_view), non_neg_integer()}
                           ]
                         )
```
**for all chain types**

I can't decide which is better, wdyt?

**Review by sl1depengwyn — APPROVED** [review]

File: apps/explorer/lib/explorer/chain/import/runner/zilliqa/aggregate_quorum_certificates.ex [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  @impl Import.Runner
+  def timeout, do: @timeout
+
+  @spec insert(Repo.t(), [map()], %{required(:timeout) => timeout(), required(:timestamps) => Import.timestamps()}) ::
+          {:ok, [AggregateQuorumCertificate.t()]}
+          | {:error, [Changeset.t()]}
+  def insert(repo, changes_list, %{timeout: timeout, timestamps: timestamps} = _options) when is_list(changes_list) do
+    # Enforce Zilliqa.AggregateQuorumCertificate ShareLocks order (see docs: sharelock.md)
```
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/import/runner/zilliqa/aggregate_quorum_certificates.ex:62]: Fixed! Please check https://github.com/blockscout/docs/pull/329.

File: .github/workflows/publish-docker-image-for-zilliqa.yml
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply .github/workflows/publish-docker-image-for-zilliqa.yml:1]: ✅

File: .github/workflows/publish-docker-image-for-zilliqa.yml [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        uses: ./.github/actions/setup-repo
+        id: setup
+        with:
+          docker-username: ${{ secrets.DOCKER_USERNAME }}
+          docker-password: ${{ secrets.DOCKER_PASSWORD }}
+          docker-remote-multi-platform: true
+          docker-arm-host: ${{ secrets.ARM_RUNNER_HOSTNAME }}
+          docker-arm-host-key: ${{ secrets.ARM_RUNNER_KEY }}
+      
+      - name: Build and push Docker image (indexer + API)
```
Comment (vbaranov) [inline .github/workflows/publish-docker-image-for-zilliqa.yml:27]: Separate images for indexer and API seems missing.

Comment (fedor-ivn) [reply .github/workflows/publish-docker-image-for-zilliqa.yml:27]: My bad. Fixed in https://github.com/blockscout/blockscout/pull/10699/commits/16ec68ec6ac87bce552c9912b6080210957c0fc3.

**Review by vbaranov — APPROVED** [review]

**Pull request merged**